### PR TITLE
Add benchmarks for FieldManager handling

### PR DESCRIPTION
**Pull request by apelisse**
We don't have a lot of data on allocations and how much time it takes to
run apply or update on objects, so adding some benchmark will help us
investigate possible improvements.

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/assign @jennybuckley @lavalamp

### CODE REVIEW

**Comment by apelisse** [comment]
cc @kwiesmueller

**Comment by lavalamp** [comment]
Can you put the current output from these benchmarks somewhere, so we have a baseline for future improvement?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go [hunk 1]
```diff
@@ -173,3 +175,86 @@ func TestApplyDoesNotStripLabels(t *testing.T) {
 		t.Fatalf("labels shouldn't get stripped on apply: %v", m)
 	}
 }
+
+func BenchmarkApplyNewObject(b *testing.B) {
+	f := NewTestFieldManager()
+
+	obj := &corev1.Pod{}
+
+	b.ReportAllocs()
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: Probably doesn't matter on these timescales, but technically a b.Reset() might be good to exclude the time it takes to construct `f`. (below, too)

**Comment by apelisse** [comment]
I'd rather suggest running the benchmarks again, because results can vary based on machine used, etc (not for allocations).

But here you go:
```
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager
BenchmarkApplyNewObject-12     	    5000	    323482 ns/op	   76018 B/op	     900 allocs/op
BenchmarkUpdateNewObject-12    	   10000	    195308 ns/op	   63540 B/op	     637 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager	3.708s
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal
BenchmarkSetToFields-12    	  100000	     21602 ns/op	    8974 B/op	     124 allocs/op
BenchmarkFieldsToSet-12    	   50000	     28240 ns/op	   12897 B/op	     147 allocs/op
BenchmarkYAMLToTyped-12    	   20000	     61974 ns/op	   25718 B/op	     253 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal	6.561s
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go (hunk 1 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: It doesn't hurt to have it ..

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go [hunk 2]
```diff
@@ -173,3 +175,86 @@ func TestApplyDoesNotStripLabels(t *testing.T) {
 		t.Fatalf("labels shouldn't get stripped on apply: %v", m)
 	}
 }
+
+func BenchmarkApplyNewObject(b *testing.B) {
+	f := NewTestFieldManager()
+
+	obj := &corev1.Pod{}
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: Would it be better to make a new object each time? I don't think we would hit any caching, but maybe better to stay safe?

Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: Yes, and this is free anyway so ...

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: Probably doesn't matter, and it doesn't seem feasible for the other tests, so ignore this :)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/fields_test.go [hunk 3]
```diff
@@ -107,3 +108,52 @@ func TestSetToFieldsError(t *testing.T) {
 		}
 	}
 }
+
+func BenchmarkSetToFields(b *testing.B) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/fields_test.go]: For these I think we should be sure to test both deep and broad sets, but we can add that later.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/typeconverter_test.go [hunk 4]
```diff
@@ -175,3 +176,59 @@ Final object:
 %#v`, obj, newObj)
 	}
 }
+
+var result typed.TypedValue
+
+func BenchmarkYAMLToTyped(b *testing.B) {
+	y := `
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/typeconverter_test.go]: We might as well use the really huge one you showed me earlier? :)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go [hunk 5]
```diff
@@ -173,3 +176,266 @@ func TestApplyDoesNotStripLabels(t *testing.T) {
...
+								"value": true
+							},
+						},
+					},
+				},
+			},
+		},
+
+	}`
+	newObj := &unstructured.Unstructured{Object: map[string]interface{}{}}
```
Comment (jennybuckley) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go]: Can this be into a corev1.Pod{} ?

**Pull request merged**